File: apps/python/design-pattern/src/observer.py
"""Observer パターン

Employee がサブジェクト（観察対象）として、
給与やタイトルの変更をオブザーバーに通知する。
"""

from typing import Protocol


class Observer(Protocol):
    """オブザーバープロトコル"""

    def update(self, employee: "Employee") -> None: ...


class Employee:
    """従業員クラス（Observer パターン - Subject）"""

    def __init__(self, name: str, title: str, salary: int) -> None:
        self.name = name
        self._title = title
        self._salary = salary
        self._observers: list[Observer] = []

    def add_observer(self, observer: Observer) -> None:
        self._observers.append(observer)

    def remove_observer(self, observer: Observer) -> None:
        self._observers.remove(observer)

    def _notify_observers(self) -> None:
        for observer in self._observers:
            observer.update(self)

    @property
    def salary(self) -> int:
        return self._salary

    @salary.setter
    def salary(self, new_salary: int) -> None:
        self._salary = new_salary
        self._notify_observers()

    @property
    def title(self) -> str:
        return self._title

    @title.setter
    def title(self, new_title: str) -> None:
        self._title = new_title
        self._notify_observers()


class Payroll:
    """給与計算オブザーバー"""

    def __init__(self) -> None:
        self.last_notification: str = ""

    def update(self, employee: Employee) -> None:
        self.last_notification = (
            f"{employee.name} の給与が {employee.salary} に変更されました"
        )


class TaxMan:
    """税務担当オブザーバー"""

    def __init__(self) -> None:
        self.last_notification: str = ""

    def update(self, employee: Employee) -> None:
        self.last_notification = (
            f"{employee.name} に新しい税金の請求書を送付します"
        )

File: apps/scala/design-pattern/src/Observer.scala
// Observer パターン
// trait と mutable リストでイベント通知を実現する

package designpattern.observer

import scala.collection.mutable.ListBuffer

trait Observer:
  def update(employee: Employee): Unit

class Employee(val name: String, private var _salary: Double):
  private val observers: ListBuffer[Observer] = ListBuffer.empty

  def salary: Double = _salary

  def salary_=(newSalary: Double): Unit =
    _salary = newSalary
    notifyObservers()

  def addObserver(observer: Observer): Unit =
    observers += observer

  def removeObserver(observer: Observer): Unit =
    observers -= observer

  def notifyObservers(): Unit =
    observers.foreach(_.update(this))

class Payroll extends Observer:
  var lastEmployee: Option[Employee] = None

  override def update(employee: Employee): Unit =
    lastEmployee = Some(employee)
    // 給与計算のために記録

class TaxMan extends Observer:
  var lastEmployee: Option[Employee] = None

  override def update(employee: Employee): Unit =
    lastEmployee = Some(employee)
    // 税金計算のために記録
